### ISSUE TITLE
Terraform fails to plan unless provider version is specified

### ISSUE THREAD

**Issue by pshima** [hashicorp/terraform#15860]
### Terraform Version
% terraform -v
Terraform v0.10.2

### Terraform Configuration Files
```hcl
provider "aws" {
  region  = "us-east-1"
}

resource "aws_security_group" "consul" {
  name   = "consul"

  ingress {
    protocol    = "-1"
    from_port   = 0
    to_port     = 0
    cidr_blocks = ["0.0.0.0/0"]
  }

  egress {
    protocol    = "-1"
    from_port   = 0
    to_port     = 0
    cidr_blocks = ["0.0.0.0/0"]
  }
}
```

### Debug Output
https://gist.github.com/pshima/ef426700b57cbfa665b38bb086a6ee84

### Panic Output
NA

### Expected Behavior
Terraform init should have downloaded the aws provider

### Actual Behavior
Terraform plan output fails

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Create new path with tf contents above
1. `terraform init`
1. `terraform apply`
1. See error output

### Important Factoids
NA

### References
Is this a bug?  The documentation says that provider constraints aren't required. https://www.terraform.io/docs/configuration/providers.html#provider-versions, it also notes that when `terraform init` is ran without provider constraints it will display a warning but this also did not occur.

Not sure if this is a bug or not, but if it is not then please update the provider example in the docs to note that provider version is required.  If I am just doing something wrong let me know.

My issue was resolved by changing the provider block to 
```
provider "aws" {
  version = "~> 0.1.4"
  region  = "us-east-1"
}
```

**Comment by apparentlymart** [comment]
Hi @pshima! Sorry this is not working properly.

The _symptom_ here looks like a problem we had when plugins were built with a newer version of the plugin system than Terraform itself, but since we addressed that in 0.10.1 I expect this is something different.

I think it would be helpful to see the exact output from `terraform init` here, since it sounds like something strange is happening.

Is it possible that you already have a build of the AWS plugin installed locally from an earlier version of Terraform? If Terraform finds a file that looks like a plugin in one of its search paths (same directory as `terraform` itself, the `~/.terraform.d/plugins` directory, specified in the `~/.terraformrc` file, and a couple more) then it will assume that you wish to use this version in preference to auto-installing, which might be problematic if the plugin it's finding is built for an older Terraform version.

**Comment by pshima** [comment]
That was indeed it, old plugins in /Users/pete/bin/terraform-provider* from back in May.  Wiping those resolved it.

Thanks!
